bellman/src/domain/arith.rs

```diff
@@ -0,0 +1,123 @@
+use std::ops;
+
+use ff::{ScalarEngine, Field, PrimeField};
+use group::CurveProjective;
+
+pub trait Group<'a,E>: Sized 
+    + Copy 
+    + Clone 
+    + Send 
+    + Sync 
+    + ops::MulAssign<&'a E::Fr>
+    + ops::SubAssign<&'a Self>
+    + ops::AddAssign<&'a Self>
+where
+    Self: 'a,
+    E: ScalarEngine    
+{
+    fn zero() -> Self;
+}
+
+pub struct Point<G>(pub G);
+
+impl<'a,G> ops::MulAssign<&'a G::Scalar> for Point<G> 
+where
+    G: CurveProjective
+{
+    fn mul_assign(&mut self, rhs: &'a G::Scalar) {
+        self.0.mul_assign(rhs.into_repr());
+    }
+} 
+
+impl<'a,G> ops::AddAssign<&'a Self> for Point<G> 
+where
+    G: CurveProjective
+{
+    fn add_assign(&mut self, rhs: &'a Self) {
+        self.0.add_assign(&rhs.0);   
+    }
+}
+
+impl<'a,G> ops::SubAssign<&'a Self> for Point<G> 
+where
+    G: CurveProjective
+{
+    fn sub_assign(&mut self, rhs: &'a Self) {
+        self.0.sub_assign(&rhs.0);
+    }
+}
+
+impl<G: CurveProjective> PartialEq for Point<G> {
+    fn eq(&self, other: &Point<G>) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<G: CurveProjective> Copy for Point<G> {}
```

Comment:
Is this incomplete

Reply:
Copy doesn't have any required methods except that the type parameter G is can also be copied.

This is rusts way of enforcing explicit vs implicit deep clone, and I will move it across to a macro.

Reply:
Resolved